### PR TITLE
fix: race conditions

### PR DESCRIPTION
**Pull request by QuentinPerez**
Depends #10

### CODE REVIEW

**Comment by rs** [comment]
How do you reproduce?

**Comment by QuentinPerez** [comment]
Hi @rs

```
go test -v -race -cpu=1,2,3,4  ./...
WARNING: DATA RACE
Read by goroutine 28:
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist.func1()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:109 +0x63

Previous write by goroutine 27:
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:96 +0x45b
  github.com/rs/rest-layer/rest.TestRoutePathParentsExists()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:205 +0x14b7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 28 (running) created at:
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:117 +0xa0c
  github.com/rs/rest-layer/rest.TestRoutePathParentsExists()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:205 +0x14b7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 27 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:515 +0x11d
  main.main()
      github.com/rs/rest-layer/rest/_test/_testmain.go:108 +0x210



WARNING: DATA RACE
Read by goroutine 29:
  github.com/rs/rest-layer/rest.(*mockHandler).Find()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:42 +0x138
  github.com/rs/rest-layer/resource.(*Resource).Find()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/resource/resource.go:165 +0x180
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist.func1()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:109 +0x124

Previous write by goroutine 28:
  github.com/rs/rest-layer/rest.(*mockHandler).Find()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:42 +0x240
  github.com/rs/rest-layer/resource.(*Resource).Find()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/resource/resource.go:165 +0x180
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist.func1()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:109 +0x124

Goroutine 29 (running) created at:
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:117 +0xa06
  github.com/rs/rest-layer/rest.TestRoutePathParentsExists()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:205 +0x14b7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 28 (finished) created at:
  github.com/rs/rest-layer/rest.ResourcePath.ParentsExist()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/resource_path.go:117 +0xa06
  github.com/rs/rest-layer/rest.TestRoutePathParentsExists()
      /Users/quentinperez/go/src/github.com/rs/rest-layer/rest/routing_test.go:205 +0x14b7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc
```

File: resource/resource.go [hunk 1]
```diff
@@ -158,6 +160,8 @@ func (r *Resource) Find(ctx context.Context, lookup *Lookup, page, perPage int)
 	if r.storage == nil {
 		return nil, ErrNoStorage
 	}
+	r.lockResources.Lock()
+	defer r.lockResources.Unlock()
```
Comment (rs) [inline resource/resource.go]: Are you sure this mutex is needed? I don't see how parallel `Find` would generate a race.

File: rest/resource_path.go [hunk 2]
```diff
@@ -109,9 +114,9 @@ func (p ResourcePath) ParentsExist(ctx context.Context) error {
 			} else {
 				c <- nil
 			}
-		}()
+		}(i)
 		// Push the resource field=value for the next hops
-		q = append(q, schema.Equal{Field: rp.Field, Value: rp.Value})
+		q = append(q, schema.Equal{Field: p[i].Field, Value: p[i].Value})
```
Comment (rs) [inline rest/resource_path.go]: Ok for this one, good catch. Would consider a separate PR? It would be awesome.

**Comment by QuentinPerez** [comment]
Yes I will close this PR and open two dedicated PR ☺️

File: resource/resource.go (hunk 1 above)
Comment (QuentinPerez) [inline resource/resource.go]: Sorry `git add .` this lines should not be added

**Pull request closed without merging**